Reuse buffer based on the size.

petro-rudenko opened the pull request:
@yuvaldeg

petro-rudenko committed:
Reuse buffer based on the size.

yuvaldeg commented on src/main/java/com/ibm/disni/util/MemBuf.java:
```diff
@@ -26,21 +26,17 @@
-	public MemBuf(MemoryAllocation.MemType type, ByteBuffer buffer,
-			long address, MemoryAllocation memAlloc, String classname) {
-		this.type = type;
+  public static long getAddress(ByteBuffer buffer) {
```
Indentation should match the other methods

yuvaldeg commented on src/main/java/com/ibm/disni/util/MemoryAllocation.java:
```diff
@@ -21,67 +21,98 @@
+		private final ConcurrentLinkedDeque<MemBuf> stack = new ConcurrentLinkedDeque<>();
+		private final int size;
+
+
```
redundant newline

yuvaldeg commented on src/main/java/com/ibm/disni/util/MemoryAllocation.java:
```diff
@@ -21,67 +21,98 @@
+		}
+
+
+		private MemBuf get() {
```
redundant newline

yuvaldeg commented on src/main/java/com/ibm/disni/util/MemoryAllocation.java:
```diff
@@ -21,67 +21,98 @@
-			buffer.order(ByteOrder.nativeOrder());
-			long address = MemoryUtils.getAddress(buffer);
-			return new MemBuf(MemType.DIRECT, buffer, address, this, classname);
+	public void put(MemBuf buf){
```
You should put a cap on the maximum idle buffers in the pool, you can make it configurable and set some default. If the cap is met, then it should free the least used pool.

yuvaldeg commented:
Please see comments inline

Commit:
SVCPostSend/Recv: Expose set API for more flexibility

Allow users with more flexibility on using and reusing the
"SVCPostSend", "SVCPostRecv" without free/allocate memory unnecessarily

Signed-off-by: Vu Pham <vu@mellanox.com>

yuvaldeg commented on src/main/java/com/ibm/disni/rdma/verbs/impl/RdmaCmNat.java:
```diff
@@ -127,14 +127,14 @@ public int bindAddr(RdmaCmId id, SocketAddress address)
         }
         int ret = nativeDispatcher._bindAddr(idPriv.getObjId(), sockBuf.address());
-        sockBuf.free();
+		memAlloc.put(sockBuf);
```
Should be spaces and not tabs, like the rest of the file

yuvaldeg commented on src/main/java/com/ibm/disni/rdma/verbs/impl/RdmaCmNat.java:
```diff
@@ -157,15 +157,15 @@ public int resolveAddr(RdmaCmId id, SocketAddress source,
         int ret = nativeDispatcher._resolveAddr(idPriv.getObjId(), 0, dstBuf.address(), timeout);
         logger.info("resolveAddr, addres " + destination.toString());
-        dstBuf.free();
+		memAlloc.put(dstBuf);
```
tabs

yuvaldeg commented on src/main/java/com/ibm/disni/rdma/verbs/impl/NatPollCqCall.java:
```diff
@@ -75,10 +75,10 @@ public void set(IbvCQ cq, IbvWC[] wcList, int ne) {
 		this.csize = wcList.length*IbvWC.CSIZE;
 		if (cmd != null){
-			cmd.free();
+			memAlloc.put(cmd);
```
why put it back if you are going to allocate the same size anyway?

yuvaldeg commented on src/main/java/com/ibm/disni/rdma/verbs/impl/NatRegMrCall.java:
```diff
@@ -61,12 +61,12 @@ public void set(IbvPd pd, long address, int length, int access) {
+
 		if (cmd != null){
-			cmd.free();
+			memAlloc.put(cmd);
```
why put it back if you are going to allocate the same size anyway?

yuvaldeg commented on src/main/java/com/ibm/disni/util/MemoryAllocation.java:
```diff
@@ -21,67 +21,142 @@
 
+public class MemoryAllocation {
+	private static final int MIN_BLOCK_SIZE = 64; // 64B
+	private int MAX_CACHE_SIZE = 4 * 1024 * 1024; // 4MB
```
sounds very small - does it get thrashed often? I'm worried about resource hungry applications

yuvaldeg commented on src/main/java/com/ibm/disni/util/MemoryAllocation.java:
```diff
@@ -21,67 +21,142 @@
-				buf.getBuffer().clear();
-				return buf;
+
+	private class AllocatorStack {
```
Please add a comment to explain what this is and how it is used - just a 2-3 lines summary

yuvaldeg reviewed (changes requested):
Please review comments inline

petro-rudenko committed:
Merge pull request #2 from vuhuong/peter-fix-memalloc

SVCPostSend/Recv: Expose set API for more flexibility

petro-rudenko committed:
Fixes based on review.

petro-rudenko committed:
Icrease cache size + fix issue on empty stacks.

apocanow1 commented:
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
dhl, @dochang, @xiilei, @endel, @coopermaruyama, @mbouaziz, @gimdongwon, @nerdsupremacist, @RaynorZhong, @piggybox, @cisen, @bengal-rtzr, @hishis, @Dentrax, @ruancaipu, @wimgithub, @jeancabral, @MKagesawa, @earthloong, @g-harel
